Sources/SwiftFormatWhitespaceLinter/WhitespaceLinter.swift

```diff
@@ -0,0 +1,332 @@
+/// code with formatted text.
+public struct WhitespaceLinter {
+
+  /// The text of the input source code to be linted.
+  let userText: String
+
+  /// The formatted version of `userText`.
+  let formattedText: String
+
+  /// The Context object containing the DiagnosticEngine.
+  let context: Context
+
+  /// Creates a new WhitespaceLinter with the given context.
+  ///
+  /// - Parameters:
+  ///   - user: The text of the Swift source code to be linted.
+  ///   - formatted: The formatted text to compare to `user`.
+  ///   - context: The context object containing the DiagnosticEngine instance we wish to use.
+  public init(user: String, formatted: String, context: Context) {
+    self.userText = user
+    self.formattedText = formatted
+    self.context = context
+  }
+
+  /// Perform whitespace linting.
+  public func lint() {
+    var userOffset = 0
+    var formOffset = 0
+    var isFirstCharater = true
+    var lastChar: Character?
+
+    repeat {
+      let userNext = nextCharacter(offset: userOffset, data: self.userText)
+      let formNext = nextCharacter(offset: formOffset, data: self.formattedText)
+
+      // `userText` and `formattedText` should only differ in their whitespace characters.
+      if userNext.char != formNext.char {
+        fatalError("Characters do not match")
+      }
+
+      lastChar = userNext.char
+
+      compareWhitespace(
+        userOffset: userOffset,
+        isFirstCharacter: isFirstCharater,
+        userWs: userNext.whitespace,
+        formattedWs: formNext.whitespace
+      )
+
+      userOffset = userNext.offset + 1
+      formOffset = formNext.offset + 1
+      isFirstCharater = false
+    } while lastChar != nil
+  }
+
+  /// Compare the whitespace buffers between the user text and formatted text, and emit linter
+  /// errors accordingly.
+  ///
+  /// Note: properly formatted whitespace will always be some number of newlines characters,
+  /// followed by some number of spaces. e.g. "\n ", "\n\n  ", "\n", " ". The user's whitespace
+  /// could have spaces and newlines in any order. e.g. " \n ", "  \n", etc.
+  ///
+  /// - Parameters:
+  ///   - userOffset: The current non-whitespace character offset within the user text.
+  ///   - isFirstCharacter: Are we at the first character in the text?
+  ///   - userWs: The user leading whitespace buffer at the current character.
+  ///   - formattedWs: The formatted leading whitespace buffer at the current character.
+  func compareWhitespace(
+    userOffset: Int, isFirstCharacter: Bool, userWs: String, formattedWs: String
+  ) {
+    if userWs == formattedWs { return }
+
+    // e.g. "\n" -> ["", ""], and "" -> [""]
+    let userTokens = userWs.split(
+      separator: "\n", omittingEmptySubsequences: false).map(String.init)
+    let formTokens = formattedWs.split(
+      separator: "\n", omittingEmptySubsequences: false).map(String.init)
+
+    indentationError(
```

Comment:
Nit: The names of these methods make them sound a little unconditional (as I read them, they feel like names of something you would call once you know you already need to emit those errors).

Names like `checkForIndentationError`, `checkForTrailingWhitespaceError`, and so forth might be clearer.

Reply:
Done.